stan/math/rev/prob/std_normal_log_qf.hpp

```diff
@@ -0,0 +1,51 @@
+#ifndef STAN_MATH_REV_PROB_STD_NORMAL_LOG_QF_HPP
+#define STAN_MATH_REV_PROB_STD_NORMAL_LOG_QF_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/fun/constants.hpp>
+#include <stan/math/prim/prob/std_normal_log_qf.hpp>
+#include <cmath>
+
+namespace stan {
+namespace math {
+
+/**
+ * The inverse of unit normal cumulative density function evaluated
+ * at the log probability
+ *
+ * The derivative is the reciprocal of unit normal density function
+ * adjusted by the derivative of the exponentiated log input
+ *
+ * @param log_p log probability
+ * @return the unit normal inverse cdf evaluated at log_p
+ */
+inline var std_normal_log_qf(const var& log_p) {
+  auto precomp_exp = to_arena(as_array_or_scalar(exp(log_p.val())));
+  return make_callback_var(
+      std_normal_log_qf(log_p.val()), [log_p, precomp_exp](auto& vi) mutable {
+        log_p.adj() += precomp_exp * vi.adj() * SQRT_TWO_PI
+                       / std::exp(-0.5 * vi.val() * vi.val());
+      });
+}
+
+/**
+ * Return the elementwise inverse of unit normal cumulative density function.
+ *
+ * @tparam T a `var_value` with inner Eigen type
+ * @param p log probability vector
+ * @return Elementwise unit normal inverse cdf
+ */
+template <typename T, require_var_matrix_t<T>* = nullptr>
+inline auto std_normal_log_qf(const T& log_p) {
+  auto precomp_exp = to_arena(as_array_or_scalar(exp(log_p.val())));
+  return make_callback_var(
+      std_normal_log_qf(log_p.val()), [log_p, precomp_exp](auto& vi) mutable {
+        log_p.adj().array() += precomp_exp * vi.adj().array() * SQRT_TWO_PI
+                               / (-0.5 * vi.val().array().square()).exp();
```

Comment:
Do you mean I only need one of these calls if I can get the `as_array_or_scalar` to work? Or do I keep the first which is for `var` type and update the second?

If I update the second (starting on row 39), is this what you were thinking
```cpp
template <typename T, require_var_matrix_t<T>* = nullptr>
inline auto std_normal_log_qf(const T& log_p) {
  return make_callback_var(
      std_normal_log_qf(log_p.val()), [log_p](auto& vi) mutable {
      const auto& vi_array = as_array_or_scalar(vi);
      const auto& deriv = as_array_or_scalar(log_p).val() 
                            + log(vi_array.adj()) 
                            - NEG_LOG_SQRT_TWO_PI 
                            + 0.5 * square(vi_array.val());
        as_array_or_scalar(log_p).adj() += exp(deriv);
      });
}
```
I get a bunch of errors regarding `vi` with the first being
```cpp
./stan/math/rev/prob/std_normal_log_qf.hpp:52:50: error: no matching function for call to ‘as_array_or_scalar(stan::math::internal::callback_vari<Eigen::Matrix<double, -1, -1>, stan::math::std_normal_log_qf<stan::math::var_value<Eigen::Matrix<double, -1, -1> >, 0>::<lambda(auto:627&)> >&)’
   52 |         const auto& vi_array = as_array_or_scalar(vi);
      |                                ~~~~~~~~~~~~~~~~~~^~~~
```

Reply:
I was about to post on how to fix that, but then I realised that I'd led you down a tricky path! Once that's fixed, there are test failures with the `rev` header because of this part: `log(vi_array.adj())`. There's no guarantee that the adjoint will be positive, so taking the log causes `-nan` values to propagate.

Sorry for the bad advice, you may need to also revert the `fwd` header changes

Reply:
do you mean the `rev` header changes?